app/helpers/application_helper.rb

```diff
@@ -316,20 +316,33 @@ def link_to_chart(name, values)
     link_to icon_tag('signal'), url, target: :blank
   end
 
-  # show which stages this reference is deploy(ed+ing) to
+  # Show which stages this reference has been or is currently being deployed to.
   def deployed_or_running_list(stages, reference)
-    html = "".html_safe
-    stages.each do |stage|
-      next unless deploy = stage.deployed_or_running_deploy
-      next unless deploy.references?(reference)
-      label = (deploy.active? ? "label-warning" : "label-success")
-
-      text = "".html_safe
-      text << stage.name
-      html << content_tag(:span, text, class: "label #{label} release-stage")
-      html << " "
+    pieces = stages.map do |stage|
+      # The first deploy is the most recent one.
+      deploy = stage.deploys.where(reference: reference).first
```

Comment:
we should fetch all the deploys in a single query with `stage.project.deploys.select('max(id) as id').where(stage_id: stages.map(&:id), reference: reference).group(:stage_id)` ?

Reply:
Won't that just give the Deploy id, not the full deploy?

Reply:
Okay, I've pushed an optimized version that only does 1 query per row.